OpenRA.Game/Server/OrderBuffer.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace OpenRA.Server
+{
+	public class OrderBuffer
+	{
+		const int NumberOfFrames = 20;
+		const int Interval = 1000;
+
+		Stopwatch gameTimer;
+		long nextUpdate = 0;
+
+		readonly Dictionary<int, long> timestamps = new Dictionary<int, long>();
+
+		readonly Dictionary<int, Queue<long>> deltas = new Dictionary<int, Queue<long>>();
+		int timestep;
+		int ticksPerSecond;
+
+		public void AddOrderTimestamp(int playerIndex)
+		{
+			var time = gameTimer.ElapsedMilliseconds;
+
+			timestamps[playerIndex] = time;
+
+			if (timestamps.Values.All(t => t > 0))
+			{
+				// pick a baseline
+				var baseline = timestamps[0];
```

Comment:
Did you test what happens if you add latency to the player with index 0 and leave the player at index 1 as is?